src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -396,36 +396,194 @@ OMapInnerNode::make_full_merge(omap_context_t oc, OMapNodeRef right)
+    auto mut_liter = mut->iter_idx(liter->get_offset());
+    auto mut_riter = mut->iter_idx(riter->get_offset());
+    return mut->do_balance(oc, mut_liter, mut_riter, l, r);
+  }
+  DEBUGT("balanced l {} r {} liter {} riter {}",
+    oc.t, *l, *r, liter->get_key(), riter->get_key());
+  return l->make_balanced(oc, r, *pivot_idx
+  ).si_then([liter=liter, riter=riter, l=l, r=r, oc, this](auto tuple) {
+    LOG_PREFIX(OMapInnerNode::merge_entry);
+    auto [replacement_l, replacement_r, replacement_pivot] = tuple;
+    replacement_l->init_range(l->get_begin(), replacement_pivot);
+    replacement_r->init_range(replacement_pivot, r->get_end());
+    DEBUGT("to update parent: {} {} {}",
+      oc.t, *this, *replacement_l, *replacement_r);
+    if (get_meta().depth > 2) { // l and r are inner nodes
+      auto &left = *l->template cast<OMapInnerNode>();
+      auto &right = *r->template cast<OMapInnerNode>();
+      auto &rep_left = *replacement_l->template cast<OMapInnerNode>();
+      auto &rep_right = *replacement_r->template cast<OMapInnerNode>();
+      this->adjust_copy_src_dest_on_balance(
+	oc.t, left, right, true, rep_left, rep_right);
+    }
+
+    //update operation will not cuase node overflow, so we can do it first
+    this->update_child_ptr(
+      liter.get_offset(),
+      dynamic_cast<base_child_t*>(replacement_l.get()));
+    journal_inner_update(
+      liter,
+      replacement_l->get_laddr(),
+      maybe_get_delta_buffer());
+    bool overflow = extent_will_overflow(replacement_pivot.size(),
+      std::nullopt);
+    if (!overflow) {
+      this->update_child_ptr(
+	riter.get_offset(),
+	dynamic_cast<base_child_t*>(replacement_r.get()));
+      journal_inner_remove(riter, maybe_get_delta_buffer());
+      journal_inner_insert(
+	riter,
+	replacement_r->get_laddr(),
+	replacement_pivot,
+	maybe_get_delta_buffer());
+      std::vector<laddr_t> dec_laddrs{l->get_laddr(), r->get_laddr()};
+      return dec_ref(oc, dec_laddrs
+      ).si_then([] {
+	return merge_entry_ret(
+	       interruptible::ready_future_marker{},
+	       mutation_result_t(mutation_status_t::SUCCESS,
+		 std::nullopt, std::nullopt));
+      });
+    } else {
+      DEBUGT("balanced and split {} r {} riter {}",
+	oc.t, *l, *r, riter.get_key());
+      //use remove and insert to instead of replace,
+      //remove operation will not cause node split, so we can do it first
+      this->remove_child_ptr(riter.get_offset());
+      journal_inner_remove(riter, maybe_get_delta_buffer());
+      return make_split_insert(
+	oc, riter, replacement_pivot, replacement_r
+      ).si_then([this, oc, l = l, r = r](auto mresult) {
+	std::vector<laddr_t> dec_laddrs{
+	  l->get_laddr(),
+	  r->get_laddr(),
+	  get_laddr()};
+	return dec_ref(oc, dec_laddrs
+	).si_then([mresult = std::move(mresult)] {
+	  return merge_entry_ret(
+		 interruptible::ready_future_marker{}, mresult);
+	});
+      });
+    }
+  });
+
+}
```

Comment:
It's hard to follow the actual changes based on the long differences here.

Can you help to split out a cleanup commit that only moves/cleans the logic into `do_balance()` and `do_merge()`, so that the diff of the actual changes can be more accurate?

Reply:
Yes, will do.